google/cloud/connection_options_test.cc

```diff
@@ -0,0 +1,262 @@
+  ASSERT_EQ(2, test_args.num_args);
+  ASSERT_EQ(GRPC_ARG_STRING, test_args.args[0].type);
+  ASSERT_EQ(GRPC_ARG_STRING, test_args.args[1].type);
+
+  // There is no (AFAICT) guarantee on the order of the arguments in this array,
+  // and the C types are hard to work with. Capture the arguments in a map to
+  // make it easier to work with them.
+  std::map<std::string, std::string> args;
+  for (std::size_t i = 0; i != test_args.num_args; ++i) {
+    args[test_args.args[i].key] = test_args.args[i].value.string;
+  }
+
+  EXPECT_EQ("testing-pool", args["grpc.channel_pooling_domain"]);
+  // The gRPC library adds its own version to the user-agent string, so we only
+  // check that our component appears in it.
+  EXPECT_THAT(args["grpc.primary_user_agent"],
+              StartsWith(options.user_agent_prefix()));
+}
+
+TEST(ConnectionOptionsTest, CustomBackgroundThreads) {
+  CompletionQueue cq;
+
+  auto options = TestConnectionOptions(grpc::InsecureChannelCredentials())
+                     .DisableBackgroundThreads(cq);
+  auto background = options.background_threads_factory()();
+
+  using ms = std::chrono::milliseconds;
+
+  // Schedule some work, it cannot execute because there is no thread attached.
+  auto background_thread_id = background->cq().MakeRelativeTimer(ms(0)).then(
+      [](future<StatusOr<std::chrono::system_clock::time_point>>) {
+        return std::this_thread::get_id();
+      });
+  EXPECT_NE(std::future_status::ready, background_thread_id.wait_for(ms(1)));
+
+  // Verify we can create our own threads to drain the completion queue.
+  std::thread t([&cq] { cq.Run(); });
+  EXPECT_EQ(std::future_status::ready, background_thread_id.wait_for(ms(100)));
+
+  cq.Shutdown();
+  t.join();
+}
+
+TEST(ConnectionOptionsTest, DefaultTracingComponentsNoEnvironment) {
+  EnvironmentVariableRestore restore("GOOGLE_CLOUD_CPP_ENABLE_TRACING");
+  internal::UnsetEnv("GOOGLE_CLOUD_CPP_ENABLE_TRACING");
+  auto const actual = internal::DefaultTracingComponents();
+  EXPECT_THAT(actual, ElementsAre());
+}
+
+TEST(ConnectionOptionsTest, DefaultTracingComponentsWithValue) {
+  EnvironmentVariableRestore restore("GOOGLE_CLOUD_CPP_ENABLE_TRACING");
+  internal::SetEnv("GOOGLE_CLOUD_CPP_ENABLE_TRACING", "a,b,c");
+  auto const actual = internal::DefaultTracingComponents();
+  EXPECT_THAT(actual, ElementsAre("a", "b", "c"));
+}
+
+TEST(ConnectionOptionsTest, DefaultTracingOptionsNoEnvironment) {
+  EnvironmentVariableRestore restore("GOOGLE_CLOUD_CPP_TRACING_OPTIONS");
+  internal::UnsetEnv("GOOGLE_CLOUD_CPP_TRACING_OPTIONS");
+  auto const actual = internal::DefaultTracingOptions();
+  auto const expected = TracingOptions{};
+  EXPECT_EQ(expected.single_line_mode(), actual.single_line_mode());
+  EXPECT_EQ(expected.use_short_repeated_primitives(),
+            actual.use_short_repeated_primitives());
+  EXPECT_EQ(expected.truncate_string_field_longer_than(),
+            expected.truncate_string_field_longer_than());
```

Comment:
Perhaps just `EXPECT_EQ(expected, actual)`?

Reply:
I do not think TracingOptions has a `operator==` or `operator!=` or a `PrintTo` defined, so that would not work.  Do we want to add them?